File: android/app/src/main/java/sol/vortexx/android/bots/impl/HttpBotsRepository.kt
package sol.vortexx.android.bots.impl

import io.ktor.client.call.body
import io.ktor.client.request.delete
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.http.isSuccess
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.Serializable
import sol.vortexx.android.bots.api.BotsRepository
import sol.vortexx.android.db.dao.BotDao
import sol.vortexx.android.db.entities.BotEntity
import sol.vortexx.android.net.impl.VortexHttpClient
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class HttpBotsRepository @Inject constructor(
    private val http: VortexHttpClient,
    private val dao: BotDao,
) : BotsRepository {

    override fun marketplace(): Flow<List<BotEntity>> = dao.observeMarketplace()
    override fun installed():   Flow<List<BotEntity>> = dao.observeInstalled()

    override suspend fun refreshMarketplace(): Boolean = runCatching {
        val resp = http.client.get("api/bots/marketplace")
        if (!resp.status.isSuccess()) return@runCatching false
        dao.upsertAll(resp.body<List<BotDto>>().map(BotDto::toEntity))
        true
    }.getOrDefault(false)

    override suspend fun install(id: Long): Boolean = runCatching {
        val resp = http.client.post("api/bots/$id/install")
        if (resp.status.isSuccess()) { dao.setInstalled(id, true); true } else false
    }.getOrDefault(false)

    override suspend fun uninstall(id: Long): Boolean = runCatching {
        val resp = http.client.delete("api/bots/$id/install")
        if (resp.status.isSuccess()) { dao.setInstalled(id, false); true } else false
    }.getOrDefault(false)

    @Serializable private data class BotDto(
        val id: Long, val name: String, val author: String = "",
        val short_description: String = "", val avatar_url: String? = null,
        val installed: Boolean = false, val rating: Double = 0.0,
        val install_count: Long = 0,
    ) {
        fun toEntity() = BotEntity(
            id = id, name = name, author = author,
            shortDescription = short_description, avatarUrl = avatar_url,
            installed = installed, rating = rating, installCount = install_count,
        )
    }
}

File: ios/Modules/Sources/Bots/impl/HttpBotsRepository.swift
import Foundation
import Net
import DB

public final class HttpBotsRepository: BotsRepository {
    private let http: HttpClient
    private let dao: BotDao

    public init(http: HttpClient, dao: BotDao) { self.http = http; self.dao = dao }

    public var marketplace: AsyncStream<[Bot]> { bridge(dao.observeMarketplace()) }
    public var installed:   AsyncStream<[Bot]> { bridge(dao.observeInstalled()) }

    public func refreshMarketplace() async -> Bool {
        do {
            let list = try await http.send(.get("api/bots/marketplace"), [BotDto].self)
            try await dao.upsertAll(list.map(\.toRecord))
            return true
        } catch { return false }
    }

    public func install(_ id: Int64) async -> Bool {
        do {
            _ = try await http.send(HttpRequest(method: .POST, path: "api/bots/\(id)/install"))
            try await dao.setInstalled(id, true); return true
        } catch { return false }
    }

    public func uninstall(_ id: Int64) async -> Bool {
        do {
            _ = try await http.send(.delete("api/bots/\(id)/install"))
            try await dao.setInstalled(id, false); return true
        } catch { return false }
    }

    private func bridge(_ src: AsyncStream<[BotRecord]>) -> AsyncStream<[Bot]> {
        AsyncStream { cont in
            let inner = Task {
                for await rows in src { cont.yield(rows.map(\.toDomain)) }
            }
            cont.onTermination = { @Sendable _ in inner.cancel() }
        }
    }

    private struct BotDto: Decodable {
        let id: Int64
        let name: String
        let author: String?
        let short_description: String?
        let avatar_url: String?
        let installed: Bool?
        let rating: Double?
        let install_count: Int64?
        var toRecord: BotRecord {
            BotRecord(id: id, name: name, author: author ?? "",
                      shortDescription: short_description ?? "",
                      avatarUrl: avatar_url,
                      installed: installed ?? false,
                      rating: rating ?? 0,
                      installCount: install_count ?? 0)
        }
    }
}

private extension BotRecord {
    var toDomain: Bot {
        Bot(id: id, name: name, author: author,
            shortDescription: shortDescription, avatarUrl: avatarUrl,
            installed: installed, rating: rating, installCount: installCount)
    }
}
